File: R/levels.R

setMethod("droplevels", signature(x="SpatRaster"),
	function(x, level=NULL, layer=1) {
		if (is.null(level)) {
			x@cpp <- x@cpp$droplevels()
			messages(x)
		} else {
			if (is.character(layer)) {
				layer <- match(layer, names(x))
				if (any(is.na(layer))) {
					error("droplevels", "invalid layer")
				}
			}
			x[[layer]][x[[layer]] %in%  level] <- NA
			x@cpp <- x@cpp$droplevels()
			messages(x)
		}
	}
)


setMethod("is.factor", signature(x="SpatRaster"),
	function(x) {
		x@cpp$hasCategories()
	}
)

setMethod("as.factor", signature(x="SpatRaster"),
	function(x) {
		x@cpp = x@cpp$makeCategorical(-1, spatOptions())
		messages(x)
		#if (!hasValues(x)) {
		#	error("as.factor", "x has no values")
		#}
		#x <- round(x)
		#u <- unique(x, TRUE)
		#for (i in 1:nlyr(x)) {
		#	set.cats(x, i, data.frame(ID=u[[i]], label=u[[i]], stringsAsFactors=FALSE))
		#}
		#x
	}
)


setMethod("levels", signature(x="SpatRaster"),
	function(x) {
		x <- x@cpp$getCategories()
		lapply(x, function(i) {
			d <- .getSpatDF(i$df)
			if (ncol(d) == 0) return("")
			d[, c(1, max(1, i$index+1))]
		})
	}
)


setMethod("levels<-", signature(x="SpatRaster"),
	function(x, value) {
		x@cpp <- x@cpp$deepcopy()
		if (is.null(value)) {
			x@cpp$removeCategories(-1)
			return(messages(x, "levels<-"))
		} else if (inherits(value, "list")) {
			for (i in 1:length(value)) {
				set.cats(x, i, value[[i]])
			}
		} else {
			set.cats(x, 1, value)
		}
		x
	}
)



setMethod ("set.cats" , "SpatRaster",
	function(x, layer=1, value, active=1) {

		if (missing(value)) {
			error("set.cats", "value cannot be missing")
			#return(invisible(x@cpp$setCatIndex(layer-1, index)))
		}

		if (is.character(layer)) {
			layer <- match(layer, names(x))
			if (any(is.na(layer))) {
				error("set.cats", "invalid layer")
			}
		}
		layer <- round(layer)

		if (length(layer) > 1) {
			if (!is.list(value)) {
				error("set.cats", "value should be a list")
			}
			if (length(layer) != length(value)) {
				error("set.cats", "length(value) != length(value)")
			}
			index <- rep_len(active, nlyr(x))
			for (i in 1:length(layer)) {
				ok <- set.cats(x, layer[i], value[[i]], index[i])
				x <- messages(x, "set.cats")
			}
			return(invisible(ok))
		} 

		if (layer < 1) {
			if (!is.list(value)) {
				error("set.cats", "value should be a list")
			}
			if (length(value) != nlyr(x)) {
				error("set.cats", "length(value) != nlyr(x)")
			}
			index <- rep_len(active, nlyr(x))
			for (i in 1:length(value)) {
				ok <- set.cats(x, i, value[[i]], index[i])
				x <- messages(x, "set.cats")
			}
			return(invisible(ok))
		}

		layer <- layer[1]
		if (is.character(layer)) {
			i <- match(layer, names(x))[1]
			if (length(i) == 0) {
				error("set.cats", layer, " is not in names(x)")
			}
			layer <- i
		} else {
			stopifnot(layer > 0 && layer <= nlyr(x))
		}

		if (inherits(value, "list")) {
			value <- value[[1]]
		}
		if (is.null(value)) {
			x@cpp$removeCategories(layer-1)
			messages(x, "set.cats")
			return(invisible(TRUE))
		}

		setname <- FALSE
		if (!is.data.frame(value)) {
			if (is.vector(value) || is.factor(value)) {
				if ((length(value) == 1) && value[1] == "") {
					return(invisible(""))
				}
				warn("set.cats", "setting categories like this is deprecated; use a two-column data.frame instead")
				value <- data.frame(value=0:(length(value)-1), category=value, stringsAsFactors=FALSE)
			} else {
				error("set.cats", "value should be a data.frame")
			}
		} else {
			setname <- TRUE
			if (ncol(value) == 1) {
				error("set.cats", "value should have at least two columns")
			} else {
				if (!is.numeric(value[[1]])) {
					error("set.cats", "the first column of 'value' must be numeric")
				}
				value[,1] <- round(value[[1]])
				if (length(unique(value[[1]])) != nrow(value)) {
					error("set.cats", "duplicate values (IDs) supplied")
				}
			}
		}
		value[[1]] <- as.integer(value[[1]])
		for (i in seq_along(value)) {
			if (is.factor(value[[i]])) {
				value[[i]] <- as.character(value[[i]])
			}
		}

		index <- max(1, min(ncol(value), active))
		if (setname) {
			nms <- names(x)
			cn <- colnames(value)[index+1]
			if (!(tolower(cn) %in% c("histogram", "count", "red", "green", "blue", "alpha", "opacity", "r", "g", "b", "a"))) {
				nms[layer] <- cn
				if (! x@cpp$setNames(nms, FALSE)) {
					error("names<-", "cannot set name")
				}
			}
		}
		if (any(is.na(value[[1]]))) {
			error("set.cats", "you cannot associate a category with NA")
		}
		if (any(table(value[[1]]) > 1)) {
			error("set.cats", "you cannot have duplicate IDs")
		}

		value <- .makeSpatDF(value)
		ok <- x@cpp$setCategories(layer-1, value, index)
		x <- messages(x, "set.cats")
		invisible(ok)
	}
)



setMethod ("categories" , "SpatRaster",
	function(x, layer=1, value, active=1, ...) {
		#... to accept but ignore old argument "index"
		x@cpp <- x@cpp$deepcopy()
		set.cats(x, layer, value, active)
		x
	}
)


setMethod ("activeCat" , "SpatRaster",
	function(x, layer=1) {
		layer <- layer[1]
		if (is.character(layer)) {
			layer = which(layer == names(x))[1]
			if (is.na(layer)) {
				error("activeCat", "invalid layer name")
			}
		}
		if (layer < 1) {
			sapply(1:nlyr(x), function(i) x@cpp$getCatIndex(i-1))
		} else {
			if (!is.factor(x)[layer]) {
				return(NA)
			}
			x@cpp$getCatIndex(layer-1)
		}
	}
)

setMethod("activeCat<-" , "SpatRaster",
	function(x, layer=1, value) {
		if (missing(value)) {
			value <- layer[1]
			layer <- 1
		} else {
			layer <- layer[1]
		}
		if ((layer < 1) | (layer > nlyr(x))) {
			error("activeCat", "invalid layer")
		}
		if (!is.factor(x)[layer]) {
			error("activeCat", "layer is not categorical")
		}
		if (is.character(value)) {
			g <- cats(x)[[layer]]
			value <- which(value == names(g))[1] - 1
			if (is.na(value)) {
				error("activeCat", "invalid category name")
			}
		}
		x <- deepcopy(x)
		if (!x@cpp$setCatIndex(layer-1, value)) {
			error("activeCat", "invalid category index")
		}
		x
	}
)

setMethod("cats" , "SpatRaster",
	function(x, layer) {
		if (!missing(layer)) {
			x <- subset(x, layer, NSE=FALSE)
		}
		cats <- x@cpp$getCategories()
		lapply(1:nlyr(x), function(i) {
			if (cats[[i]]$df$nrow == 0) {
				return(NULL)
			}
			.getSpatDF(cats[[i]]$df)
		})
	}
)


# superseded by levels(x)[[layer]]
..active_cats <- function(x, layer) {
	ff <- is.factor(x)
	if (!any(ff)) {
		return (lapply(ff, function(i) NULL))
	}
	cats <- x@cpp$getCategories()
	x <- lapply(1:length(cats), function(i) {
		if (cats[[i]]$df$nrow == 0) return(NULL)
		r <- .getSpatDF(cats[[i]]$df)
		a <- activeCat(x, i)
		if (a < 0) return(NULL)
		r[, c(1, a+1)]
	})

	if (!missing(layer)) {
		x[[layer]]
	} else {
		x
	}
}


setMethod ("as.numeric", "SpatRaster",
	function(x, index=NULL, filename="", ...) {
		if (!any(is.factor(x))) {
			x <- deepcopy(x)
			x@cpp$setValueType(0)
			return(x)
		}
		if (nlyr(x) > 1) {
			x <- lapply(1:nlyr(x), function(i) as.numeric(x[[i]], index=index))
			x <- rast(x)
			if (filename != "") {
				x <- writeRaster(x, filename, ...)
			}
			return(x)
		}
		g <- cats(x)[[1]]
		if (!is.null(index)) {
			if (is.character(index)) {
				index <- match(index, colnames(g))
				if (is.na(index)) {
					error("as.numeric", "index is not category name")				
				}
				if (index == 1) {
					levels(x) <- NULL
					x@cpp$setValueType(0)
					if (filename != "") {
						x <- writeRaster(x, filename, ...)
					}
					return(x)
				}		
			} else {
				index <- round(index[1])
				if (!((index >= 1) & (index < ncol(g)))) {
					error("as.numeric", "index out of range")
				}
				index <- index + 1
			}
		} else {
			index <- activeCat(x, 1)
			if (index <= 1) {
				levels(x) <- NULL
				x@cpp$setValueType(0)
				if (filename != "") {
					x <- writeRaster(x, filename, ...)
				}
				return(x)
			}
		}
		from <- g[,1]
		to <- g[,index]
		if (!is.numeric(to)) {
			suppressWarnings(toto <- as.numeric(to))
			if (sum(is.na(toto) > sum(is.na(to)))) {
				to <- as.integer(as.factor(to))
			} else {
				to <- toto
			}
		}
		m <- cbind(from, to)
		m <- m[!is.na(m[,1]), ,drop=FALSE]
		classify(x, m, names=names(g)[index], filename, ...)
	}
)



catLayer <- function(x, index, ...) {
		stopifnot(nlyr(x) == 1)
		if (!is.factor(x)) return(x)
		g <- cats(x)[[1]]
		if (!is.null(index)) {
			if (!((index > 1) & (index <= ncol(g)))) {
				error("as.numeric", "invalid index")
			}
		} else {
			index <- set.cats(x, 1)
		}
		from <- g[,1]
		toc <- g[,index]

		addFact <- FALSE
		if (!is.numeric(toc)) {
			addFact <- TRUE
			ton <- as.integer(as.factor(toc))
		} else {
			ton <- toc
		}
		m <- cbind(from, ton)
		m <- m[!is.na(m[,1]), ,drop=FALSE]
		x <- classify(x, m, names=names(g)[index], ...)
		if (addFact) {
			fact <- unique(data.frame(ton, toc))
			names(fact) <- c("ID", names(g)[index])
			fact <- fact[order(fact[,1]), ]
			set.cats(x, 1, fact)
		}
		x
}



setMethod("catalyze", "SpatRaster",
	function(x, filename="", ...) {
		g <- cats(x)
		out <- list()
		for (i in 1:nlyr(x)) {
			y <- x[[i]]
			gg <- g[[i]]
			if (nrow(gg) > 0) {
				for (j in 2:ncol(gg)) {
					z <- as.numeric(y, index=j-1)
					out <- c(out, z)
				}
			} else {
				out <- c(out, y)
			}
		}
		out <- rast(out)
		if (filename!="") {
			out <- writeRaster(out, filename, ...)
		}
		out
	}
)



setMethod("concats", "SpatRaster",
	function(x, y, filename="", ...) {
		opt <- spatOptions(filename, ...)
		x@cpp = x@cpp$combineCats(y@cpp, opt)
		messages(x, "concats")
	}
)

setMethod("addCats", "SpatRaster",
	function(x, value, merge=FALSE, layer=1) {
		if (!(is.factor(x)[layer])) {
			error("addCat", "the layer has no categories to add to")
		}
		cts <- cats(x)[[layer]]
		nact <- ncol(cts)
		if (merge) {
			if (ncol(value) < 2) {
				error("addCat", "'value' must have at least two columns when using 'merge=TRUE'")
			}
			cts <- merge(cts, value, by=1, all.x=TRUE)
			cts <- cts[order(cts[,1]), ]
		} else {
			if (nrow(cts) != nrow(value)) {
				error("addCat", "the number of categories does not match")
			}
			cts <- cbind(cts, value)
		}
		categories(x, layer=layer, cts, active=nact)
	}
)

File: python/src/terra/levels.py
"""
levels.py — categorical raster levels (factors), color tables.
"""
from __future__ import annotations
from typing import Optional, Union, List, Dict, Any

from ._terra import SpatRaster, SpatOptions
from ._helpers import _getSpatDF, _makeSpatDF, messages
from .names import set_names_inplace


def _opt() -> SpatOptions:
    return SpatOptions()


def _layer_idx(x: SpatRaster, layer: Union[int, str]) -> int:
    """0-based layer index (matches C++ ``source`` layer indexing)."""
    if isinstance(layer, str):
        return list(x.names).index(layer)
    return int(layer)


# ---------------------------------------------------------------------------
# is_factor / as_factor
# ---------------------------------------------------------------------------

def is_factor(x: SpatRaster) -> List[bool]:
    """Return per-layer flags indicating whether the layer is categorical."""
    return list(x.hasCategories())


def as_factor(x: SpatRaster) -> SpatRaster:
    """Convert *x* to a categorical raster (in-place on a copy)."""
    x = x.deepcopy() if hasattr(x, 'deepcopy') else x
    x.pntr = x.makeCategorical(-1, _opt())
    return messages(x, "as_factor")


# ---------------------------------------------------------------------------
# levels / set levels
# ---------------------------------------------------------------------------

def levels(x: SpatRaster, layer: Optional[Union[int, str]] = None) -> List[Any]:
    """
    Return the categorical lookup table(s) for *x*.

    Parameters
    ----------
    x : SpatRaster
    layer : int or str, optional
        If given, return only the table for that layer.

    Returns
    -------
    list of pandas.DataFrames (or None for non-factor layers).
    """
    cats = x.getCategories()
    result = []
    for cat in cats:
        df = _getSpatDF(cat.df)
        if df is None or len(df.columns) == 0:
            result.append(None)
        else:
            idx = cat.index + 1
            cols = [df.columns[0], df.columns[min(idx, len(df.columns) - 1)]]
            result.append(df[cols])
    if layer is not None:
        if isinstance(layer, str):
            nms = list(x.names)
            layer = nms.index(layer)
        return result[layer]
    return result


def set_levels(x: SpatRaster, value: Union[List, Any], active: int = 1) -> SpatRaster:
    """
    Set the categorical lookup table for *x*.

    Parameters
    ----------
    x : SpatRaster
    value : DataFrame or list of DataFrames
        Lookup table(s) with columns [value_id, label, ...].
    active : int
        Which column (1-based) is the active label column.

    Returns
    -------
    SpatRaster
    """
    xc = x.deepcopy() if hasattr(x, 'deepcopy') else x
    if not isinstance(value, list):
        value = [value]
    for i, v in enumerate(value):
        if v is None:
            xc.removeCategories(i)
        else:
            set_cats(xc, i, v, active)
    return xc


# ---------------------------------------------------------------------------
# cats / set_cats / categories
# ---------------------------------------------------------------------------

def cats(x: SpatRaster, layer: Optional[Union[int, str]] = None) -> List[Any]:
    """
    Return the full category data frame(s) for *x*.

    Parameters
    ----------
    x : SpatRaster
    layer : int or str, optional
        If given, return only for that layer.

    Returns
    -------
    list of pandas.DataFrames (or None).
    """
    raw = x.getCategories()
    result = []
    for cat in raw:
        if cat.df.nrow == 0:
            result.append(None)
        else:
            result.append(_getSpatDF(cat.df))
    if layer is not None:
        if isinstance(layer, str):
            nms = list(x.names)
            layer = nms.index(layer)
        return result[layer]
    return result


def set_cats(
    x: SpatRaster,
    layer: Union[int, str] = 0,
    value: Any = None,
    active: int = 1,
) -> bool:
    """
    Set the category data frame for a single layer of *x* (in-place).

    Parameters
    ----------
    x : SpatRaster
    layer : int or str
        Layer index (0-based) or name.
    value : pandas.DataFrame
        Lookup table with at least two columns: integer IDs and labels.
    active : int
        Which label column (1-based offset past the ID column) is active.

    Returns
    -------
    bool
    """
    layer_idx = _layer_idx(x, layer)

    if value is None:
        x.removeCategories(layer_idx)
        return True

    import pandas as pd
    if not isinstance(value, pd.DataFrame):
        raise TypeError("value must be a pandas DataFrame")
    if len(value.columns) < 2:
        raise ValueError("value must have at least two columns")
    # Ensure first col is int
    value = value.copy()
    value.iloc[:, 0] = value.iloc[:, 0].astype(int)
    sdf = _makeSpatDF(value)
    idx = max(1, min(len(value.columns) - 1, active))
    ok = x.setCategories(layer_idx, sdf, idx)
    return bool(ok)


def categories(
    x: SpatRaster,
    layer: Union[int, str] = 0,
    value: Any = None,
    active: int = 1,
) -> SpatRaster:
    """
    Return a copy of *x* with updated category table for *layer*.

    Parameters
    ----------
    x : SpatRaster
    layer : int or str
    value : pandas.DataFrame
    active : int

    Returns
    -------
    SpatRaster
    """
    xc = x.deepcopy() if hasattr(x, 'deepcopy') else x
    set_cats(xc, layer, value, active)
    return xc


# ---------------------------------------------------------------------------
# active_cat
# ---------------------------------------------------------------------------

def active_cat(x: SpatRaster, layer: Union[int, str] = 0) -> int:
    """
    Return the index of the active label column for a categorical layer.

    Parameters
    ----------
    x : SpatRaster
    layer : int or str
        Layer (0-based index or name).

    Returns
    -------
    int
        Column index within the category table (C++ convention).
    """
    return int(x.getCatIndex(_layer_idx(x, layer)))


def set_active_cat(
    x: SpatRaster,
    value: Union[int, str],
    layer: Union[int, str] = 0,
) -> SpatRaster:
    """
    Return a copy of *x* with a different active category column.

    Parameters
    ----------
    x : SpatRaster
    value : int or str
        Column index or column name to activate.
    layer : int or str
        Layer (0-based index or name).

    Returns
    -------
    SpatRaster
    """
    layer_idx = _layer_idx(x, layer)

    if isinstance(value, str):
        ct = cats(x, layer_idx)
        if ct is None:
            raise ValueError("layer is not categorical")
        col_idx = list(ct.columns).index(value)
    else:
        col_idx = int(value)

    xc = x.deepcopy() if hasattr(x, 'deepcopy') else x
    if not xc.setCatIndex(layer_idx, col_idx):
        raise RuntimeError("invalid category index")
    return xc


# ---------------------------------------------------------------------------
# add_cats / concats / catalyze / drop_levels
# ---------------------------------------------------------------------------

def add_cats(
    x: SpatRaster,
    value: Any,
    merge: bool = False,
    layer: Union[int, str] = 0,
) -> SpatRaster:
    """
    Add columns to the category table of a layer.

    Parameters
    ----------
    x : SpatRaster
    value : pandas.DataFrame
        New column(s) to add; first column must be the ID.
    merge : bool
        If True, merge by ID rather than binding columns.
    layer : int or str

    Returns
    -------
    SpatRaster
    """
    import pandas as pd

    li = _layer_idx(x, layer)
    cts = cats(x, li)
    if cts is None:
        raise ValueError("layer has no categories to add to")
    nact = len(cts.columns)
    if merge:
        if len(value.columns) < 2:
            raise ValueError("value must have at least two columns when merge=True")
        cts = pd.merge(cts, value, on=cts.columns[0], how="left")
        cts = cts.sort_values(cts.columns[0]).reset_index(drop=True)
    else:
        if len(cts) != len(value):
            raise ValueError("number of categories does not match")
        cts = pd.concat([cts, value.reset_index(drop=True)], axis=1)
    return categories(x, layer=li, value=cts, active=nact)


def drop_levels(x: SpatRaster) -> SpatRaster:
    """Drop unused levels from all categorical layers."""
    xc = x.droplevels()
    return messages(xc, "drop_levels")


def concats(x: SpatRaster, y: SpatRaster) -> SpatRaster:
    """Combine the category tables of *x* and *y*."""
    opt = _opt()
    xc = x.combineCats(y, opt)
    return messages(xc, "concats")


def catalyze(x: SpatRaster) -> SpatRaster:
    """Convert categorical column values to separate numeric layers."""
    ct = cats(x)
    if not any(c is not None for c in ct):
        return x
    from_vals = []
    to_vals_list = []
    nms = None
    for i, c in enumerate(ct):
        if c is not None and len(c) > 0:
            nms = list(c.columns[1:])
            from_vals = c.iloc[:, 0].tolist()
            for j in range(1, len(c.columns)):
                col = c.iloc[:, j]
                try:
                    to_vals_list.append([float(v) for v in col])
                except (ValueError, TypeError):
                    import pandas as pd
                    to_vals_list.append([float(i) for i in pd.Categorical(col).codes + 1])
            break
    if not from_vals:
        return x
    opt = _opt()
    xc = x.lookup_catalyze(from_vals, to_vals_list, opt)
    result = messages(xc, "catalyze")
    if nms:
        set_names_inplace(result, [str(n) for n in nms])
    return result


# ---------------------------------------------------------------------------
# color tables
# ---------------------------------------------------------------------------

def has_colors(x: SpatRaster) -> List[bool]:
    """Return per-layer flags indicating whether a color table exists."""
    return list(x.hasColors())


def coltab(x: SpatRaster) -> List[Any]:
    """
    Return the color table(s) for *x*.

    Returns
    -------
    list of pandas.DataFrames (or None for layers without color tables).
    """
    has = list(x.hasColors())
    if not any(has):
        return [None] * len(has)
    raw = x.getColors()
    result = []
    for i, df_raw in enumerate(raw):
        if has[i]:
            result.append(_getSpatDF(df_raw))
        else:
            result.append(None)
    return result


def set_coltab(
    x: SpatRaster,
    value: Any,
    layer: Union[int, str] = 0,
) -> SpatRaster:
    """
    Set a color table for a layer of *x*.

    Parameters
    ----------
    x : SpatRaster
    value : list of str (hex colors), array, or DataFrame with RGBA columns.
    layer : int or str

    Returns
    -------
    SpatRaster
    """
    import pandas as pd
    import numpy as np

    layer_idx = _layer_idx(x, layer)

    xc = x.deepcopy() if hasattr(x, 'deepcopy') else x

    if value is None:
        xc.removeColors(layer_idx)
        return xc

    if isinstance(value, (list, np.ndarray)) and not isinstance(value, pd.DataFrame):
        value_list = list(value)
        if len(value_list) > 0 and isinstance(value_list[0], str):
            # Parse hex strings
            rows = []
            for idx, hex_color in enumerate(value_list):
                hex_color = hex_color.lstrip('#')
                if len(hex_color) == 6:
                    r, g, b = int(hex_color[0:2], 16), int(hex_color[2:4], 16), int(hex_color[4:6], 16)
                    a = 255
                elif len(hex_color) == 8:
                    r, g, b, a = (int(hex_color[i:i+2], 16) for i in range(0, 8, 2))
                else:
                    r, g, b, a = 128, 128, 128, 255
                rows.append([idx, r, g, b, a])
            value = pd.DataFrame(rows, columns=["value", "red", "green", "blue", "alpha"])

    if isinstance(value, pd.DataFrame):
        value = value.copy()
        value.iloc[:, 0] = value.iloc[:, 0].astype(int)
        for col in value.columns[1:]:
            value[col] = value[col].clip(0, 255).astype(int)
        value = value.fillna(255)
        sdf = _makeSpatDF(value)
        if not xc.setColors(layer_idx, sdf):
            raise RuntimeError("could not set color table")
    return xc
